### ISSUE TITLE
Throwing exception with autofac bootstrapper

### ISSUE THREAD

**Issue by abhijit838** [SimpleAuthentication/SimpleAuthentication#217]
## **Dependency installed and IAuthenticationCallbackProvider implemented or not exeption is same as below:**
System.InvalidOperationException occurred
  HResult=0x80131509
  Message=Something went wrong when trying to satisfy one of the dependencies during composition, make sure that you've registered all new dependencies in the container and inspect the innerexception for more details.
  Source=Nancy
  StackTrace:
   at Nancy.Bootstrapper.NancyBootstrapperBase`1.SafeGetNancyEngineInstance()
   at Nancy.Bootstrapper.NancyBootstrapperBase`1.GetEngine()
   at Nancy.Owin.NancyMiddleware.UseNancy(NancyOptions options)
   at Nancy.Owin.DelegateExtensions.UseNancy(Action`1 builder, NancyOptions options)
   at Juna.Zone.ZoneRestApi.Startup.<>c.<Configure>b__1_0(Action`1 x) in C:\Users\plank-abhijit\Documents\Visual Studio 2017\Projects\zone-backend-restapi\Juna.Zone.RestApi\Startup.cs:line 15
   at Microsoft.AspNetCore.Builder.OwinExtensions.UseOwin(IApplicationBuilder builder, Action`1 pipeline)
   at Juna.Zone.ZoneRestApi.Startup.Configure(IApplicationBuilder app) in C:\Users\plank-abhijit\Documents\Visual Studio 2017\Projects\zone-backend-restapi\Juna.Zone.RestApi\Startup.cs:line 15

Inner Exception 1:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = NancyEngine (ReflectionActivator), Services = [Nancy.INancyEngine], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 2:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = DefaultRequestDispatcher (ReflectionActivator), Services = [Nancy.Routing.IRequestDispatcher], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 3:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = DefaultRouteResolver (ReflectionActivator), Services = [Nancy.Routing.IRouteResolver], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 4:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = RouteCache (ReflectionActivator), Services = [Nancy.Routing.IRouteCache], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 5:
DependencyResolutionException: An exception was thrown while invoking the constructor 'Void .ctor(Nancy.INancyModuleCatalog, Nancy.INancyContextFactory, Nancy.Routing.IRouteSegmentExtractor, Nancy.Routing.IRouteDescriptionProvider, Nancy.Culture.ICultureService, System.Collections.Generic.IEnumerable`1[Nancy.Routing.IRouteMetadataProvider])' on type 'RouteCache'.

Inner Exception 6:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = INancyModule[] (DelegateActivator), Services = [System.Collections.Generic.IEnumerable`1[[Nancy.INancyModule, Nancy, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null]]], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = ExternallyOwned

Inner Exception 7:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = INancyModule[] (DelegateActivator), Services = [System.Collections.Generic.IEnumerable`1[[Nancy.INancyModule, Nancy, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null]]], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = ExternallyOwned

Inner Exception 8:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = SimpleAuthenticationModule (ReflectionActivator), Services = [Nancy.INancyModule], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = OwnedByLifetimeScope

Inner Exception 9:
DependencyResolutionException: None of the constructors found with 'Autofac.Core.Activators.Reflection.DefaultConstructorFinder' on type 'Nancy.SimpleAuthentication.SimpleAuthenticationModule' can be invoked with the available services and parameters:
Cannot resolve parameter 'Nancy.SimpleAuthentication.IAuthenticationCallbackProvider callbackProvider' of constructor 'Void .ctor(Nancy.SimpleAuthentication.IAuthenticationCallbackProvider)'.
Cannot resolve parameter 'Nancy.SimpleAuthentication.IAuthenticationCallbackProvider callbackProvider' of constructor 'Void .ctor(Nancy.SimpleAuthentication.IAuthenticationCallbackProvider, SimpleAuthentication.Core.IConfigurationOptions)'.

## **After Implemention and class registered in Autofac bootstrapper like below:**
`container.Update(builder =>
			{
builder.Register(c => new SocialAuthenticationProvider()).As<IAuthenticationCallbackProvider>();
}`

**Getting the below exception:** 
System.InvalidOperationException occurred
  HResult=0x80131509
  Message=Something went wrong when trying to satisfy one of the dependencies during composition, make sure that you've registered all new dependencies in the container and inspect the innerexception for more details.
  Source=Nancy
  StackTrace:
   at Nancy.Bootstrapper.NancyBootstrapperBase`1.SafeGetNancyEngineInstance()
   at Nancy.Bootstrapper.NancyBootstrapperBase`1.GetEngine()
   at Nancy.Owin.NancyMiddleware.UseNancy(NancyOptions options)
   at Nancy.Owin.DelegateExtensions.UseNancy(Action`1 builder, NancyOptions options)
   at Juna.Zone.ZoneRestApi.Startup.<>c.<Configure>b__1_0(Action`1 x) in C:\Users\plank-abhijit\Documents\Visual Studio 2017\Projects\zone-backend-restapi\Juna.Zone.RestApi\Startup.cs:line 15
   at Microsoft.AspNetCore.Builder.OwinExtensions.UseOwin(IApplicationBuilder builder, Action`1 pipeline)
   at Juna.Zone.ZoneRestApi.Startup.Configure(IApplicationBuilder app) in C:\Users\plank-abhijit\Documents\Visual Studio 2017\Projects\zone-backend-restapi\Juna.Zone.RestApi\Startup.cs:line 15

Inner Exception 1:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = NancyEngine (ReflectionActivator), Services = [Nancy.INancyEngine], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 2:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = DefaultRequestDispatcher (ReflectionActivator), Services = [Nancy.Routing.IRequestDispatcher], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 3:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = DefaultRouteResolver (ReflectionActivator), Services = [Nancy.Routing.IRouteResolver], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 4:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = RouteCache (ReflectionActivator), Services = [Nancy.Routing.IRouteCache], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope

Inner Exception 5:
DependencyResolutionException: An exception was thrown while invoking the constructor 'Void .ctor(Nancy.INancyModuleCatalog, Nancy.INancyContextFactory, Nancy.Routing.IRouteSegmentExtractor, Nancy.Routing.IRouteDescriptionProvider, Nancy.Culture.ICultureService, System.Collections.Generic.IEnumerable`1[Nancy.Routing.IRouteMetadataProvider])' on type 'RouteCache'.

Inner Exception 6:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = INancyModule[] (DelegateActivator), Services = [System.Collections.Generic.IEnumerable`1[[Nancy.INancyModule, Nancy, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null]]], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = ExternallyOwned

Inner Exception 7:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = INancyModule[] (DelegateActivator), Services = [System.Collections.Generic.IEnumerable`1[[Nancy.INancyModule, Nancy, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null]]], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = ExternallyOwned

Inner Exception 8:
DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = SimpleAuthenticationModule (ReflectionActivator), Services = [Nancy.INancyModule], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = OwnedByLifetimeScope

Inner Exception 9:
DependencyResolutionException: An exception was thrown while invoking the constructor 'Void .ctor(Nancy.SimpleAuthentication.IAuthenticationCallbackProvider)' on type 'SimpleAuthenticationModule'.

Inner Exception 10:
MissingMethodException: Method not found: 'RouteBuilder Nancy.NancyModule.get_Get()'.

**Comment by PureKrome** [comment]
Hi @abhijit838  - thanks for posting the issue and looking at/using SimpleAuthentication :)

Can you please provide a bit more code for me to help.

Can u show me some code to how u're doing all/any SimpleAuthentication registration. I saw soe stuff in the middle of the huge stack trace error.

also, what version of Nancy ? 
what version of SimpleAuthentication ?

etc..

**Comment by abhijit838** [comment]
Thanks for reply.
 I made it work. The problem with the registration instance references. I had to register 3 classes - 
`builder.Register(c => new SocialAuthenticationCallbackProvider()).As<IAuthenticationCallbackProvider>().AsSelf().InstancePerLifetimeScope();
				builder.Register(c => new ConfigurationOptions()).As<IConfigurationOptions>().AsSelf().InstancePerLifetimeScope();
				builder.Register(c => new SessionCache(new Session())).As<ICache>().AsSelf().InstancePerLifetimeScope();`

And also as I am using Nancy 2.0.0-clinteastwood for that I had remove the before pipeline from SimpleAuthenticationModule.
